x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/RequestXContent.java

```diff
@@ -114,82 +126,121 @@ private static ObjectParser<EsqlQueryRequest, Void> objectParserAsync(Supplier<E
         return parser;
     }
 
-    private static List<TypedParamValue> parseParams(XContentParser p) throws IOException {
-        List<TypedParamValue> result = new ArrayList<>();
+    private static QueryParams parseParams(XContentParser p) throws IOException {
+        List<QueryParam> result = new ArrayList<>();
+        List<String> errors = new ArrayList<>();
         XContentParser.Token token = p.currentToken();
+        boolean namedParameter = false;
+        boolean unnamedParameter = false;
 
         if (token == XContentParser.Token.START_ARRAY) {
             Object value = null;
-            String type = null;
-            TypedParamValue previousParam = null;
-            TypedParamValue currentParam;
+            DataType type = null;
+            QueryParam currentParam = null;
+            TempObjects param;
 
             while ((token = p.nextToken()) != XContentParser.Token.END_ARRAY) {
                 XContentLocation loc = p.getTokenLocation();
 
                 if (token == XContentParser.Token.START_OBJECT) {
                     // we are at the start of a value/type pair... hopefully
-                    currentParam = PARAM_PARSER.apply(p, null);
+                    param = PARAM_PARSER.apply(p, null);
+                    if (param.fields.size() > 1) {
+                        errors.add("Multiple parameters in one curly bracket is not supported.");
+                    }
+                    for (Map.Entry<String, Object> entry : param.fields.entrySet()) {
+                        if (isValidParamName(entry.getKey()) == false) {
+                            errors.add(
+                                entry.getKey()
+                                    + " is not a valid parameter name. "
+                                    + "A valid parameter name starts with a letter and contains letters, digits and underscores only."
+                            );
+                        }
+                        type = EsqlDataTypes.fromJava(entry.getValue());
+                        if (type == null) {
+                            errors.add(loc + " " + entry + " is not supported as a parameter.");
+                        }
+                        currentParam = new QueryParam(entry.getKey(), entry.getValue(), type);
+                        namedParameter = true;
+                        if (unnamedParameter && namedParameter) {
+                            errors.add(
+                                "Params cannot contain both named and unnamed parameters: "
```

Comment:
The style of the error message should be: what is the problem with the params (and your error message is ok), followed by something like `got [.....]` where we show what exactly are the values with problems.

With this error message in your code we have something like
`"params": [5,7,{"z":6}] Failed to parse params: [Params cannot contain both named and unnamed parameters: 6 [z][INTEGER], [5 [null][INTEGER], 7 [null][INTEGER]]]",`. I find it difficult to follow the format `6 [z] [INTEGER]`...

Some suggestions for improvements:
- ideally, the error message should look like this `Params cannot contain both named and unnamed parameters; got [{"z": 6}] and [5, 7]`
- shorten the list of params in the error message. For example, catch the first/last named param, first/last unnamed param and list only those. In my example above: `Params cannot contain both named and unnamed parameters; got [{"z": 6}] and [5]`